### ISSUE TITLE
route function does not work if route contains curly braces

### ISSUE THREAD

**Issue by sclause2412** [laravel/framework#46266]
- Laravel Version: 10.0.3
- PHP Version: 8.2.2
- Database Driver & Version: MySQL (MariaDB 10.9.5)

### Description:
If a route is generated by passing a Model the route function automatically takes the ID column of the model and places it in the route. If this ID column value contains curly braces it is taken for the route and then the route tries to replace the curly braces again and that leads to an error.

### Steps To Reproduce:
1. Create a new Model where the ID column is a string
2. Add a new entry where the ID is something with curly braces, e.g. 123{1}
3. Add a resource route for this model, e.g. Route::resource('tests', TestController::class); 
4. Add a index view where you list all models.
5. Add a route link to show a detailed page for the models by using a blade directive, e.g. {{ route('tests.show', $test) }}

This will cause an error since $test is replaced by the ID and then the URL is translated from /tests/{test} to /tests/123{1} which will then be again translated - which of course fails because there is no "element" {1}.

The same behavior happens if the link is created like this:
{{ route('tests.show', $test->id) }}
results in: /tests/{test} --> /tests/123{1} --> error

The only possibility is to use it this way:
{{ route('tests.show', urlencode($test->id)) }}
results in: /tests/{test} --> /tests/123%7B1%7D --> OK

**Comment by ankurk91** [comment]
Why would someone use `{` in their primary keys?

**Comment by sclause2412** [comment]
For easy understanding I just took the ID here. But for example you are creating a blog and you want to use the slug instead of the ID. This slug should of course be a clean and simple string without special characters and I would do it that way.
I just made a typo in a test and then I realized this problem. Even if it is not recommended it should not produce an error.

Additionally, it is only a problem because the route function does not urlencode the data and not because the link does not work. If i just type http://example.com/tests/123{1} in the browser everything works fine.
